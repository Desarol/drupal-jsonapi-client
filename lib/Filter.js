"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

/**
 * Based on: https://www.drupal.org/docs/8/modules/jsonapi/filtering
 */
var Filter =
/*#__PURE__*/
function () {
  function Filter(_ref) {
    var identifier = _ref.identifier,
        path = _ref.path,
        _ref$operator = _ref.operator,
        operator = _ref$operator === void 0 ? '=' : _ref$operator,
        value = _ref.value,
        memberOf = _ref.memberOf;

    _classCallCheck(this, Filter);

    this.identifier = identifier;
    this.path = path;
    this.operator = operator;
    this.value = value;
    this.memberOf = memberOf;
  }
  /* eslint-disable prefer-template */


  _createClass(Filter, [{
    key: "query",
    value: function query() {
      var _this = this;

      return ["filter[".concat(this.identifier, "][condition][path]=").concat(this.path), this.operator !== '=' ? "filter[".concat(this.identifier, "][condition][operator]=").concat(this.operator) : '', this.memberOf ? "filter[".concat(this.identifier, "][condition][memberOf]=").concat(this.memberOf) : '', typeof this.value.map === 'function' ? this.value.map(function (singleValue) {
        return "filter[".concat(_this.identifier, "][condition][value][]=").concat(singleValue);
      }) : "filter[".concat(this.identifier, "][condition][value]=").concat(this.value)];
    }
    /* eslint-enable prefer-template */

  }]);

  return Filter;
}();

exports.default = Filter;