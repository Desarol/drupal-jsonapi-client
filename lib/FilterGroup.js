"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

function _toConsumableArray(arr) { return _arrayWithoutHoles(arr) || _iterableToArray(arr) || _nonIterableSpread(); }

function _nonIterableSpread() { throw new TypeError("Invalid attempt to spread non-iterable instance"); }

function _iterableToArray(iter) { if (Symbol.iterator in Object(iter) || Object.prototype.toString.call(iter) === "[object Arguments]") return Array.from(iter); }

function _arrayWithoutHoles(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = new Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

/**
 * Based on: https://www.drupal.org/docs/8/modules/jsonapi/filtering
 */
var FilterGroup =
/*#__PURE__*/
function () {
  function FilterGroup(_ref) {
    var _this = this;

    var identifier = _ref.identifier,
        type = _ref.type,
        memberOf = _ref.memberOf,
        children = _ref.children;

    _classCallCheck(this, FilterGroup);

    this.identifier = identifier;
    this.type = type;
    this.memberOf = memberOf;
    this.children = children.map(function (child) {
      // eslint-disable-next-line no-param-reassign
      child.memberOf = _this.identifier;
      return child;
    });
  }
  /* eslint-disable prefer-template */


  _createClass(FilterGroup, [{
    key: "query",
    value: function query() {
      return ["filter[".concat(this.identifier, "][group][conjunction]=").concat(this.type), this.memberOf ? "filter[".concat(this.identifier, "][group][memberOf]=").concat(this.memberOf) : ''].concat(_toConsumableArray(this.children.map(function (child) {
        return child.query();
      })));
    }
    /* eslint-enable prefer-template */

  }]);

  return FilterGroup;
}();

exports.default = FilterGroup;