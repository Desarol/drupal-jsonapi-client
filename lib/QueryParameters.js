"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

/**
 * Array.prototype.flat() is not supported in IE.
 * https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/flat
 */

/* eslint-disable */
if (!Array.prototype.flat) {
  Array.prototype.flat = function (depth) {
    var flattend = [];

    (function flat(array, depth) {
      var _iteratorNormalCompletion = true;
      var _didIteratorError = false;
      var _iteratorError = undefined;

      try {
        for (var _iterator = array[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
          var el = _step.value;

          if (Array.isArray(el) && depth > 0) {
            flat(el, depth - 1);
          } else {
            flattend.push(el);
          }
        }
      } catch (err) {
        _didIteratorError = true;
        _iteratorError = err;
      } finally {
        try {
          if (!_iteratorNormalCompletion && _iterator.return != null) {
            _iterator.return();
          }
        } finally {
          if (_didIteratorError) {
            throw _iteratorError;
          }
        }
      }
    })(this, Math.floor(depth) || 1);

    return flattend;
  };
}
/* eslint-enable */


var QueryParameters =
/*#__PURE__*/
function () {
  function QueryParameters(queryParameters) {
    _classCallCheck(this, QueryParameters);

    this.queryParameters = queryParameters;
  }

  _createClass(QueryParameters, [{
    key: "toString",
    value: function toString(depth) {
      return this.queryParameters.flat(depth).map(function (item) {
        return !!item && item.query ? item.query() : item;
      }).flat(depth).filter(function (item) {
        return !!item;
      }).join('&');
    }
  }]);

  return QueryParameters;
}();

exports.default = QueryParameters;