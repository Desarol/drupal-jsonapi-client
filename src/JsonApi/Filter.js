/**
 * Based on: https://www.drupal.org/docs/8/modules/jsonapi/filtering
 */

export default class Filter {
  constructor({
    identifier,
    path,
    operator = '=',
    value,
    memberOf,
  }) {
    this.identifier = identifier
    this.path = path
    this.operator = operator
    this.value = value
    this.memberOf = memberOf
  }

  /* eslint-disable prefer-template */
  query() {
    return ([
      `filter[${this.identifier}][condition][path]=${this.path}`,
      (
        this.operator !== '='
          ? `filter[${this.identifier}][condition][operator]=${this.operator}`
          : ''
      ),
      (
        this.memberOf
          ? `filter[${this.identifier}][condition][memberOf]=${this.memberOf}`
          : ''
      ),
      (
        typeof this.value.map === 'function'
          ? this.value.map(singleValue => `filter[${this.identifier}][condition][value][]=${singleValue}`)
          : `filter[${this.identifier}][condition][value]=${this.value}`
      ),
    ])
  }
  /* eslint-enable prefer-template */
}
