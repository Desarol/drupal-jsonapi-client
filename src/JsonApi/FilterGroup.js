/**
 * Based on: https://www.drupal.org/docs/8/modules/jsonapi/filtering
 */

export default class FilterGroup {
  constructor({
    identifier,
    type,
    memberOf,
    children,
  }) {
    this.identifier = identifier
    this.type = type
    this.memberOf = memberOf
    this.children = children.map((child) => {
      // eslint-disable-next-line no-param-reassign
      child.memberOf = this.identifier
      return child
    })
  }

  /* eslint-disable prefer-template */
  query() {
    return ([
      `filter[${this.identifier}][group][conjunction]=${this.type}`,
      (this.memberOf ? `filter[${this.identifier}][group][memberOf]=${this.memberOf}` : ''),
      ...this.children.map(child => child.query()),
    ])
  }
  /* eslint-enable prefer-template */
}
