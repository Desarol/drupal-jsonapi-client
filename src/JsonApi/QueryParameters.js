/**
 * Array.prototype.flat() is not supported in IE.
 * https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/flat
 */

/* eslint-disable */
if (!Array.prototype.flat) {
  Array.prototype.flat = function(depth) {
    var flattend = [];
    (function flat(array, depth) {
      for (let el of array) {
        if (Array.isArray(el) && depth > 0) {
          flat(el, depth - 1); 
        } else {
          flattend.push(el);
        }
      }
    })(this, Math.floor(depth) || 1);
    return flattend;
  };
}
/* eslint-enable */

export default class QueryParameters {
  constructor(queryParameters) {
    this.queryParameters = queryParameters
  }

  toString(depth) {
    return this.queryParameters
      .flat(depth)
      .map(item => (!!item && item.query ? item.query() : item))
      .flat(depth)
      .filter(item => !!item)
      .join('&')
  }
}
