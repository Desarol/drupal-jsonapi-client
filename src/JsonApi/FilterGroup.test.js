/* eslint-disable import/first */
require('regenerator-runtime/runtime') // eslint-disable-line

import QueryParameters from './QueryParameters'
import FilterGroup from './FilterGroup'
import Filter from './Filter'

describe('Filter', () => {
  it('serializes filter group with conjunction AND to a query', () => {
    expect((new QueryParameters([
      new FilterGroup({
        identifier: 'group-1',
        type: 'AND',
        children: [
          new Filter({
            identifier: 'filter-1',
            path: 'title',
            value: 'asdf123',
          }),
          new Filter({
            identifier: 'filter-2',
            path: 'status',
            value: 1,
          }),
        ],
      }),
    ])).toString(10)).toMatchSnapshot()
  })

  it('serializes filter group with conjunction OR to a query', () => {
    expect((new QueryParameters([
      new FilterGroup({
        identifier: 'group-1',
        type: 'OR',
        children: [
          new Filter({
            identifier: 'filter-1',
            path: 'title',
            value: 'asdf123',
          }),
          new Filter({
            identifier: 'filter-2',
            path: 'status',
            value: 1,
          }),
        ],
      }),
    ])).toString(10)).toMatchSnapshot()
  })

  it('serializes nested filter groups', () => {
    expect((new QueryParameters([
      new FilterGroup({
        identifier: 'group-3',
        type: 'OR',
        children: [
          new FilterGroup({
            identifier: 'group-1',
            type: 'AND',
            children: [
              new Filter({
                identifier: 'filter-1',
                path: 'title',
                value: 'asdf123',
              }),
              new Filter({
                identifier: 'filter-2',
                path: 'status',
                value: 1,
              }),
            ],
          }),
          new FilterGroup({
            identifier: 'group-2',
            type: 'AND',
            children: [
              new Filter({
                identifier: 'filter-3',
                path: 'title',
                value: 'asdf123',
              }),
              new Filter({
                identifier: 'filter-4',
                path: 'status',
                value: 1,
              }),
            ],
          }),
        ],
      }),
    ])).toString(10)).toMatchSnapshot()
  })
})
