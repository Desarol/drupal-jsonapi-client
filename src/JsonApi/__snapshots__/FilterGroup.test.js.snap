// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filter serializes filter group with conjunction AND to a query 1`] = `"filter[group-1][group][conjunction]=AND&filter[filter-1][condition][path]=title&filter[filter-1][condition][memberOf]=group-1&filter[filter-1][condition][value]=asdf123&filter[filter-2][condition][path]=status&filter[filter-2][condition][memberOf]=group-1&filter[filter-2][condition][value]=1"`;

exports[`Filter serializes filter group with conjunction OR to a query 1`] = `"filter[group-1][group][conjunction]=OR&filter[filter-1][condition][path]=title&filter[filter-1][condition][memberOf]=group-1&filter[filter-1][condition][value]=asdf123&filter[filter-2][condition][path]=status&filter[filter-2][condition][memberOf]=group-1&filter[filter-2][condition][value]=1"`;

exports[`Filter serializes nested filter groups 1`] = `"filter[group-3][group][conjunction]=OR&filter[group-1][group][conjunction]=AND&filter[group-1][group][memberOf]=group-3&filter[filter-1][condition][path]=title&filter[filter-1][condition][memberOf]=group-1&filter[filter-1][condition][value]=asdf123&filter[filter-2][condition][path]=status&filter[filter-2][condition][memberOf]=group-1&filter[filter-2][condition][value]=1&filter[group-2][group][conjunction]=AND&filter[group-2][group][memberOf]=group-3&filter[filter-3][condition][path]=title&filter[filter-3][condition][memberOf]=group-2&filter[filter-3][condition][value]=asdf123&filter[filter-4][condition][path]=status&filter[filter-4][condition][memberOf]=group-2&filter[filter-4][condition][value]=1"`;
